rep-0151.rst

```diff
@@ -0,0 +1,151 @@
+gets increasingly high the further up in the stack the package is.
+
+Updating all ROS packages which contain Python code to support Python 2
+as well as 3 is a significant effort.
+Therefore it is expected that this will take a while and will require the ROS
+community to share the necessary workload.
+
+Proposal
+========
+
+For a smooth transition the following timeline is being proposed:
+
+ * ROS O-turtle (May 2020, LTS):
+
+   * The only Debian packages provided are using Python 3.
+   * From-source builds are using Python 3 by default.
+   * All ROS packages should support Python 3.
+   * Python 2 support is not required anymore.
+
+ * ROS N-turtle (May 2019):
+
+   * The set of recommended Debian packages is using Python 3.
+   * There is an alternative set of Debian packages which is using Python 2,
+     this can be used in case some packages don't support Python 3 yet.
+   * From-source builds are using Python 3 by default, optionally choosing
+     Python 2 should be possible for all packages.
+   * All ROS packages should support Python 2 as well as Python 3.
+
+ * ROS M-turtle (May 2018, LTS):
+
+   * The set of recommended Debian packages is using Python 2.
+   * There is an alternative set of Debian packages which is using Python 3,
+     this can be used by the CI infrastructure as well as users choosing to use
+     Python 3.
+   * From-source builds are using Python 2 by default, optionally choosing
+     Python 3 should be possible.
+   * All ROS packages should support Python 2 while support for Python 3 is
+     recommended (as it has been for the past ROS distributions).
+
+At this point the REP doesn't try to specify a specific minimum minor version
+for Python 3 support but focuses on the general effort to transition to Python
+3.
+The minimum version can be defined in REP 3 [2]_ together with all other
+versions.
+
+Requirements
+============
+
+To support the above proposal a few capabilities have to be developed:
+
+ * `catkin` / `catkin_tools` need to support choosing the Python version.
+
+ * `rosdep`: choose between Python 2 and Python 3 dependencies.
+   Multiple different options can be considered for this:
+
+   * Beside the `python.yaml` file create a `python3.yaml` file and let
+     `rosdep` choose between them.
+   * Use different installers for Python 2 and Python 3 packages, e.g. `xenial`
+     and `xenial_python3`, and let `rosdep` choose between them.
+   * Create explicit rosdep keys with a `python2` and `python3` prefix and use
+     them in the package manifest files with conditional dependencies as
+     specified in REP 149 [3]_.
+
+ * Across all ROS packages which contain Python code it needs to be ensured
+   that they support both major Python versions.
+   In the majority of the cases this can be easily achieved, some examples how
+   to convert Python 2-only code to work for both major versions can be found
+   in the ROS wiki [4]_.
```

Comment:
Again, "actively-maintained packages". 

Also, "easily" is an misleading. Bilingual Python is possible, but not easy. The effort here to make it less difficult is very worthwhile.

Reply:
I actually do think that "easy" is appropriate here. The majority of necessary changes mentioned e.g. in the wiki page are trivial transformations like "if you use this than change it to that". Their are certainly cases which are more challenging though.

Reply:
The hard part is not making the changes, but doing the testing, which is absolutely necessary for Python.

Making that easier to do will help a lot.

Reply:
I completely agree. That is why this REP aims to provide an easy way to build and test with a specific Python version as well as provide support from the buildfarm to automatically run the tests with both Python versions.

Reply:
As soon as `unicode`/`str` is involved, it becomes a big pain in the butt.